pkg/pool-weighted/contracts/managed/ManagedPool.sol

```diff
@@ -277,58 +285,101 @@ contract ManagedPool is ManagedPoolSettings {
-        balanceTokenOut = _upscale(balanceTokenOut, scalingFactorTokenOut);
+        uint256 endingBalanceTokenIn;
+        uint256 endingBalanceTokenOut;
 
         if (request.kind == IVault.SwapKind.GIVEN_IN) {
             // All token amounts are upscaled.
-            request.amount = _upscale(request.amount, scalingFactorTokenIn);
+            request.amount = _upscale(request.amount, swapTokenData.scalingFactorTokenIn);
 
             // We round the amount in down (favoring a higher fee amount).
             request.amount = request.amount.mulDown(swapFeeComplement);
 
             uint256 amountOut = WeightedMath._calcOutGivenIn(
                 balanceTokenIn,
-                tokenInWeight,
+                swapTokenData.tokenInWeight,
                 balanceTokenOut,
-                tokenOutWeight,
+                swapTokenData.tokenOutWeight,
                 request.amount
             );
 
             // amountOut tokens are exiting the Pool, so we round down.
-            return _downscaleDown(amountOut, scalingFactorTokenOut);
+            amountCalculated = _downscaleDown(amountOut, swapTokenData.scalingFactorTokenOut);
         } else {
             // All token amounts are upscaled.
-            request.amount = _upscale(request.amount, scalingFactorTokenOut);
+            request.amount = _upscale(request.amount, swapTokenData.scalingFactorTokenOut);
 
             uint256 amountIn = WeightedMath._calcInGivenOut(
                 balanceTokenIn,
-                tokenInWeight,
+                swapTokenData.tokenInWeight,
                 balanceTokenOut,
-                tokenOutWeight,
+                swapTokenData.tokenOutWeight,
                 request.amount
             );
 
             // We round the amount in up (favoring a higher fee amount).
             amountIn = amountIn.divUp(swapFeeComplement);
 
             // amountIn tokens are entering the Pool, so we round up.
-            return _downscaleUp(amountIn, scalingFactorTokenIn);
+            amountCalculated = _downscaleUp(amountIn, swapTokenData.scalingFactorTokenIn);
         }
+
+        // If circuit breakers are set, check the lower bound on the tokenIn, and the upper bound on the tokenOut.
+        uint256 lowerBoundBptPrice = CircuitBreakerStorageLib.getBptPriceBound(
+            _getCircuitBreakerState(request.tokenIn),
+            swapTokenData.tokenInWeight,
+            true
+        );
+        uint256 upperBoundBptPrice = CircuitBreakerStorageLib.getBptPriceBound(
+            _getCircuitBreakerState(request.tokenOut),
+            swapTokenData.tokenOutWeight,
+            false
+        );
+        uint256 actualSupply = _getActualSupply(_getVirtualSupply());
+
+        _require(
+            !(CircuitBreakerLib.hasCircuitBreakerTripped(
+                actualSupply,
+                swapTokenData.tokenInWeight,
+                endingBalanceTokenIn,
+                lowerBoundBptPrice,
+                true // check lower bound
```

Comment:
I think i'd defintely split this into two requires, explaining we're checking the lower bound of token in (to make sure we don't buy too much of a bad thing), and the upper bound of the token out